### PR TITLE
GNSS sensor component

### PR DESCRIPTION
**Pull request by piotr-zyskowski-rai**

### CODE REVIEW

File: Code/Source/GNSS/ROS2GNSSSensorComponent.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+#include "AzCore/Math/Matrix4x4.h"
+
+#include <AzCore/Serialization/EditContext.h>
+#include <AzCore/Serialization/EditContextConstants.inl>
+
+namespace ROS2
+{
+    namespace Internal
+    {
+        const char* kGNSSMsgType = "geometry_msgs::msg::PoseStamped";
```
Comment (adamdbrw) [inline Code/Source/GNSS/ROS2GNSSSensorComponent.cpp]: Should be https://github.com/ros2/common_interfaces/blob/master/sensor_msgs/msg/NavSatFix.msg
as well as https://github.com/ros2/common_interfaces/blob/master/sensor_msgs/msg/NavSatStatus.msg

The pose can be used as well, but needs explanation on what the data is.

File: Code/Source/GNSS/GNSSFormatConversions.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    float Deg2Rad(float deg) {
+        return deg * AZ::Constants::Pi / 180.0f;
+    }
+
+    AZ::Vector3 WGS84ToECEF(const AZ::Vector3& latitudeLongitudeAltitude) {
+        const double latitudeRad = Deg2Rad(latitudeLongitudeAltitude.GetX());
+        const double longitudeRad = Deg2Rad(latitudeLongitudeAltitude.GetY());
+        const double altitude = latitudeLongitudeAltitude.GetZ();
```
Comment (adamkrawczyk) [inline Code/Source/GNSS/GNSSFormatConversions.cpp:30]: can be changed to constexpr global (used not in one function and constant)

File: Code/Source/GNSS/GNSSFormatConversions.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    AZ::Vector3 ECEFToWGS84(const AZ::Vector3& ECFEPoint) {
+        const double x = ECFEPoint.GetX();
+        const double y = ECFEPoint.GetY();
+        const double z = ECFEPoint.GetZ();
+
+        const double earthSemimajorAxis = 6378137.0f;
+        const double reciprocalFlattening = 1.0f / 298.257223563f;
+        const double earthSemiminorAxis = earthSemimajorAxis * (1.0f - reciprocalFlattening);
```
Comment (adamkrawczyk) [inline Code/Source/GNSS/GNSSFormatConversions.cpp]: to constexpr

File: Code/Source/GNSS/GNSSFormatConversions.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    //! Converts degrees to radians
+    float Deg2Rad(float deg);
+
+    //! Converts point in 1984 World Geodetic System (GS84) to Earth Centred Earth Fixed (ECEF)
+    //! @param latitudeLongitudeAltitude - point's latitude, longitude and altitude as 3d vector.
+    //!     latitude and longitude are in decimal degrees
+    //!     altitude is in meters
+    //! @return 3d vector of ECEF coordinates.
+    AZ::Vector3 WGS84ToECEF(const AZ::Vector3& latitudeLongitudeAltitude);
```
Comment (adamkrawczyk) [inline Code/Source/GNSS/GNSSFormatConversions.h:25]: What do you think about adding link to the paper/ page where mathematical formulas are taken from?
Same also in other conversions

File: Code/Tests/GNSSTest.cpp [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+
+    TEST_F(GNSSTest, ENUToECEF) {
+        const std::vector<std::tuple<AZ::Vector3, AZ::Vector3, AZ::Vector3>> input_gold_set = {
+                { {50.0f, -120.0f, -100.0f}, {-0.076833f, -0.3202f, -0.2969f}                               , { -2053900.0f, -3557459.0f, 4862712.0f}                         },
+                { {11.0f, 21.0f, 400.0f},    {-109638.9539891188f, -110428.2398398574f, -2004.501240225796f}, {5903307.167667380f, 2148628.092761247f, 1100300.642188661f}    },
+                { {-72.0f, 169.0f, 1000.0f}, {38187.58712786288f, 222803.8182465429f, -4497.428919329745f}  , {-2154856.524084172f, 379959.3447517005f, -5971509.853428957f}  },
+        };
+        for (const auto& [refWGS84, input, goldResult] : input_gold_set) {
```
Comment (adamkrawczyk) [inline Code/Tests/GNSSTest.cpp]: A changed sequence of input <-> refWGS84, everywhere else input data is first

File: Code/Source/GNSS/GNSSFormatConversions.h (hunk 4 above)
Comment (piotr-zyskowski-rai) [reply Code/Source/GNSS/GNSSFormatConversions.h:25]: There is a link in cpp. We agreed it's ehough

File: Code/Source/GNSS/ROS2GNSSSensorComponent.cpp (hunk 1 above)
Comment (piotr-zyskowski-rai) [reply Code/Source/GNSS/ROS2GNSSSensorComponent.cpp]: Changed to NavSatFix :+1:

**Review by adamkrawczyk — APPROVED** [review]

**Pull request merged**